ext/openssl/ossl_pkey.c

```diff
@@ -457,6 +457,43 @@ ossl_pkey_check_public_key(const EVP_PKEY *pkey)
     ossl_raise(ePKeyError, "public key missing");
 }
 
+/*
+ *  call-seq:
+ *      pkey.compare?(another_pkey) -> true|false
+ *
+ * Used primarily to check if an OpenSSL::X509::Certificate#public_key compares to its private key.
+ *
+ * == Example
+ *   x509 = OpenSSL::X509::Certificate.new(pem_encoded_certificate)
+ *   rsa_key = OpenSSL::PKey::RSA.new(pem_encoded_private_key)
+ *
+ *   rsa_key.compare?(x509.public_key) => true | false
+*/
+static VALUE
+ossl_pkey_compare(VALUE self, VALUE other)
+{
+    int ret;
+    EVP_PKEY *selfPKey;
+    EVP_PKEY *otherPKey;
+
+    GetPKey(self, selfPKey);
+    GetPKey(other, otherPKey);
+
+    // Explicitly check the key type given EVP_PKEY_ASN1_METHOD(3)
+    // docs param_cmp could return any negative number.
+    if (EVP_PKEY_id(selfPKey) != EVP_PKEY_id(otherPKey))
+        ossl_raise(rb_eTypeError, "cannot match different PKey types");
+
+    ret = EVP_PKEY_cmp(selfPKey, otherPKey);
+
+    if (ret == 0)
+        return Qfalse;
+    else if (ret == 1)
+        return Qtrue;
+    else
+        ossl_raise(ePKeyError, "Could not compare the keys. Please check openssl's param_cmp or pub_cmp methods implemented for the given key type.");
```

Comment:
`ossl_raise(ePKeyError, "msg")` translates to `rb_raise(ePKeyError, "msg: <details>")`, where `<details>` is an error message that OpenSSL emits. Since we don't know what caused the error here, we should just `ossl_raise(ePKeyError, "EVP_PKEY_cmp")` and let OpenSSL describe.

[EVP_PKEY_cmp() itself doesn't seem to add any message to the queue at the moment, but that's probably just because it didn't get much attention since 2006.]